### ISSUE TITLE
Experimental RBILUK 

### ISSUE THREAD

**Issue by mndevec** [trilinos/Trilinos#1401]
This issue will track the development of new RBILUK.
Goals:
 1- Clear separation of the symbolic, numeric and solve phases. 
 2- Replacing old blas calls such as GEMM, GETF2, with the register blocking little block implementations in KokkosKernels.

The issue will be developed in the fork:
https://github.com/mndevec/Trilinos/tree/develop

@trilinos/ifpack2

**Comment by mhoemmen** [comment]
@csiefer2 @brian-kelley @jhux2

**Comment by mndevec** [comment]
There are some fixes that improved the time. These changes improved run-times significantly on my desktop machine. Commit c93575c8bb11923d7fb6e399d1e02bfb8641a8bb has the below updates.

I have added a test "Ifpack2_PRBILUK.exe" associated with the source code "Ifpack2_UnitTestPerfRBILUK.cpp". 
This test reads a hard-coded matrix "laplace.mtx" in matrix market file format. I generated this using muelu's driver, with 125K rows and 860K nnz.
The test creates a block crs matrix from the read CRS Graph structure. Each block is currently 5x5. Initial run results for this test is as below:

###################################################################################
0. BILU_double_int_int_BILUIMPROVE_UnitTest ... READING
READ COMPLETE 
Initing
computing
applying
Init Time:0.289078
Compute Time:17.3222
Apply Time:0.133149
[Passed] (20 sec)
1. BILU_double_int_longlong_BILUIMPROVE_UnitTest ... READING
READ COMPLETE 
Initing
computing
applying
Init Time:0.276501
Compute Time:16.7634
Apply Time:0.299412
[Passed] (19.5 sec)

###################################################################################

Fixes: The fixes can be enabled/disabled by commenting out the line #define IFPACK2_RBILUK_IMPROVE in Ifpack2_Experimental_RBILUK_def.hpp. 

1- initAllValues were being called in the symbolic phase. This is removed out of initialize. 
2- Kokkos::deep_copy was being called for copying tiny block matrices. In threaded environments, it was using whole machine to copy the tiny matrix, causing false sharing. This is changed to a sequential copy. 
3- At the end of the loop, colFlags were being reset. However, reset loop size was the number of columns (n), rather than the number of columns in that row, this was causing O(n^2) work.
4- I replaced the Tpetra::GEMM calls, with the Kyungjoo's batched block GEMM implementation. This also slightly improved the performance.
   Kyungjoo's batched block GEMM implementation is not checked in to the trilinos. One needs to snapshot KokkosKernels develop branch as a replacement of KokkosKernels package in Trilinos to use this.
5- More TODO: I will replace the Tpetra GETRF, GETRI calls with Kyungjoo's block implementations. More GEMM replacements will follow in solve phase as well. However most improvements come from 2 and 3, and the runtime is reduced as below. 

#########################################

0. BILU_double_int_int_BILUIMPROVE_UnitTest ... READING
READ COMPLETE 
Initing
computing
applying
Init Time:0.138588
Compute Time:0.631188
Apply Time:0.139076
[Passed] (2.95 sec)

1. BILU_double_int_longlong_BILUIMPROVE_UnitTest ... READING
READ COMPLETE 
Initing
computing
applying
Init Time:0.170728
Compute Time:0.689583
Apply Time:0.138162
[Passed] (3.07 sec)
##########################################

Basically around 2x speedup on initialize, and 25x on compute phases on the input laplace matrix. 

Note: Andrew is working on a Block matrix generator that guarantees that problem can be LU factorizable. Current driver tries to make the block matrix diagonally dominant, which may not work for all problems. This works for laplace problem for now. 

@trilinos/ifpack2 @srajama1

**Comment by srajama1** [comment]
Thanks, Mehmet ! Let us integrate this once you get back and we can test it well.

**Comment by mndevec** [comment]
@bmpersc 

Let me keep this open for a bit more so that I won't forget. Once the kokkoskernels changes are fixed, I will need to convert the GEMMs to kokkoskernels implementation instead,.

**Comment by bmpersc** [comment]
@mndevec I didn't close this manually. I did, however, update the Trilinos master branch about 2 hours ago.

Looking at the commit linked the reason why this was closed is that the commit has the phrase "fixed #1401" in the commit message which is something github monitors to help automate updates to issues. The full context is "Ifpack2: RBILUK, changes of Kokkoskernels are fixed #1401." Which seems like this is saying that KokkosKernels was fixed, but github isn't smart enough to process the context.

You can find a list of words followed by an issue number that github will trigger off of to close an issue at https://help.github.com/articles/closing-issues-using-keywords/

**Comment by mhoemmen** [comment]
@trilinos/ifpack2 I think this is done.
